server/src/main/java/io/crate/replication/logical/action/ReplayChangesAction.java

```diff
@@ -86,63 +96,103 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
-            ActionListener.completeWith(
-                listener,
-                () -> new WritePrimaryResult<>(
+        protected void shardOperationOnPrimary(Request request, IndexShard primary, ActionListener<PrimaryResult<Request, ReplicationResponse>> listener) {
+            performOnPrimary(primary, request, new ArrayList<>(), 0, results -> {
+                Translog.Location location = null;
+                try {
+                    for (Engine.Result engineResult : results) {
+                        location = syncOperationResultOrThrow(engineResult, location);
+                    }
+                } catch (Exception e) {
+                    listener.onFailure(e);
+                }
+                listener.onResponse(new WritePrimaryResult<>(
                     request,
                     new ReplicationResponse(),
-                    performOnPrimary(request, primary),
+                    location,
                     null,
-                    primary
-                )
-            );
+                    primary)
+                );
+            }, listener::onFailure);
         }
 
-        public Translog.Location performOnPrimary(Request request, IndexShard primary) throws Exception {
-            Translog.Location location = null;
-
-            for (var translogOp : request.changes()) {
-                final var op = withPrimaryTerm(translogOp, primary.getOperationPrimaryTerm());
+        private void performOnPrimary(IndexShard primary, Request request, List<Engine.Result> accumulator, int offset, Consumer<List<Engine.Result>> result, Consumer<Exception> failure) {
+            for (int itemsToProcess = offset; itemsToProcess < request.changes().size(); itemsToProcess++) {
+                final var op = request.changes().get(itemsToProcess);
+                final var opWithPrimary = withPrimaryTerm(op, primary.getOperationPrimaryTerm());
                 if (primary.getMaxSeqNoOfUpdatesOrDeletes() < request.maxSeqNoOfUpdatesOrDeletes()) {
                     primary.advanceMaxSeqNoOfUpdatesOrDeletes(request.maxSeqNoOfUpdatesOrDeletes());
                 }
-                var result = primary.applyTranslogOperation(op, Engine.Operation.Origin.PRIMARY);
-                if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
-                    throw new RuntimeException("Mapping update required");
+                Engine.Result engineResult = null;
+                try {
+                    engineResult = primary.applyTranslogOperation(opWithPrimary, Engine.Operation.Origin.PRIMARY);
+                } catch (IOException e) {
+                    failure.accept(e);
+                    return;
+                }
+                if (engineResult.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED ||
+                    engineResult.getResultType() == Engine.Result.Type.FAILURE) {
```

Comment:
Is it safe to retry also on a regular failure?

Not sure if waiting for a mapping update makes sense if the mapping update is not required? Especially given that the observer filters for `isIndexMetadataUpdated`

Reply:
Ah, this catches the follow case: 
```
org.elasticsearch.index.mapper.StrictDynamicMappingException: mapping set to strict, dynamic introduction of [name] within [default] is not allowed
```
which is has the Type `Engine.Result.Type.FAILURE`.

Reply:
So we could go more specific here to avoid to catch all Failure and only this specific type.

Reply:
:+1: for handling the specific `StrictDynamicMappingException`